applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_solver.py

```diff
@@ -168,8 +196,40 @@ def InitializeSolutionStep(self):
             else:
                 self._GetLevelSetConvectionProcess().Execute()
 
+            KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Level-set convection is performed.")
+
             # Recompute the distance field according to the new level-set position
-            self._GetVariationalDistanceProcess().Execute()
+            if (self._reinitialization_type == "variational"):
+                self._GetDistanceReinitializationProcess().Execute()
+            elif (self._reinitialization_type == "parallel"):
+                adjusting_parameter = 0.05
+                layers = int(adjusting_parameter*self.main_model_part.NumberOfElements()) # this parameter is essential
+                max_distance = 1.0 # use this parameter to define the redistancing range
+                # if using CalculateInterfacePreservingDistances(), the initial interface is preserved
+                self._GetDistanceReinitializationProcess().CalculateDistances(
+                    self.main_model_part,
+                    KratosMultiphysics.DISTANCE,
+                    KratosMultiphysics.NODAL_AREA,
+                    layers,
+                    max_distance,
+                    self._GetDistanceReinitializationProcess().CALCULATE_EXACT_DISTANCES_TO_PLANE) #NOT_CALCULATE_EXACT_DISTANCES_TO_PLANE)
+
+            if (self._reinitialization_type != "none"):
+                KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Redistancing process is finished.")
+
+            # filtering noises is necessary for curvature calculation
+            if (self._distance_smoothing):
+                # distance gradient is used as a boundary condition for smoothing process
+                self._GetDistanceGradientProcess().Execute()
+                self._GetDistanceSmoothingProcess().Execute()
+                KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Smoothing process is finished.")
+
+            if (self.main_model_part.ProcessInfo[KratosCFD.SURFACE_TENSION]):
+                self._GetDistanceGradientProcess().Execute()
+                # curvature is calculated using nodal distance gradient
+                self._GetDistanceCurvatureProcess().Execute()
+                # it is needed to store level-set consistent nodal PRESSURE_GRADIENT for stabilization purpose
+                self._GetConsistentNodalPressureGradientProcess().Execute()
```

Comment:
Note that you are calling twice the `_GetDistanceGradientProcess().Execute()` if both distance smoothing and surface tension are activated.

Reply:
Smoothing changes DISTANCE, so, it is preferred to call the gradient process again.

Reply:
OK. I think we could live with the old gradients but it is a fair argument. In this case I'd put a code in the comment explaining this to avoid accidental changes in the future.

Reply:
Done.